lib/propolis/src/hw/nvme/mod.rs

```diff
@@ -822,61 +1051,111 @@ impl PciNvme {
-                    );
-                    return Err(if is_cq {
-                        NvmeError::InvalidCompQueue(qid)
-                    } else {
-                        NvmeError::InvalidSubQueue(qid)
-                    });
-                }
-
                 // 32-bit register but ignore reserved top 16-bits
                 let val = wo.read_u32() as u16;
+
                 probes::nvme_doorbell!(|| (
                     off as u64,
                     qid,
                     u8::from(is_cq),
                     val
                 ));
-                if is_cq {
-                    // Completion Queue y Head Doorbell
-                    let cq = state.get_cq(qid)?;
-                    cq.notify_head(val)?;
-                } else {
-                    // Submission Queue y Tail Doorbell
-                    let sq = state.get_sq(qid)?;
-                    sq.notify_tail(val)?;
-                }
-
-                // Poke block backend to service new requests
-                //
-                // This is done for CQs in addition to SQs, since we may have
-                // skipped pulling entries off some SQ due to CQs having no
-                // available entry slots.
-                //
-                // The state lock must be released before issuing the
-                // notification to avoid deadlocking with the block backend
-                // attempting to fetch new requests.
-                drop(state);
-                self.block_attach.notify();
+                self.ring_doorbell(qid, is_cq, val)?;
             }
         }
 
         Ok(())
     }
 
+    /// Perform the actual work of a doorbell ring
+    fn ring_doorbell(
+        &self,
+        qid: QueueId,
+        is_cq: bool,
+        val: u16,
+    ) -> Result<(), NvmeError> {
+        if !self.is_enabled.load(Ordering::Acquire) {
+            slog::debug!(
+                self.log,
+                "Doorbell write while controller is disabled"
+            );
+            return Err(if is_cq {
+                NvmeError::InvalidCompQueue(qid)
+            } else {
+                NvmeError::InvalidSubQueue(qid)
+            });
+        }
+
+        let to_notify = if is_cq {
+            // Completion Queue y Head Doorbell
+            let guard = self
+                .queues
+                .get_cq(qid)
+                .ok_or(NvmeError::InvalidCompQueue(qid))?;
+            let cq = guard.as_ref().unwrap();
+
+            cq.notify_head(val)?;
+
+            // If this CQ was previously full, SQs may have become
+            // corked while trying to get permits.  Notify them that
+            // there may now be capacity.
+            let to_notify = cq.kick();
+
+            // Query the number of entries to notify the block layer about on
+            // these now-uncorked SQIDs.
+            //
+            // Do this without holding on to the lock for the CQ slot.
+            drop(guard);
```

Comment:
Do we _need_ to hold the guard on the CQ slot over `cq.notify_head(val)?; cq.kick();`? Would it be incorrect to `Arc::clone(guard.as_ref().unwrap())`, then operate on that? I think it would only be wrong if we were calling `fire_interrupt()` and potentially sending an interrupt to the guest about it. Holding the lock avoids having to think about this entirely, of course..